File: rael/2024_3/0301/PGS_롤케이크자르기.java
import java.util.*;
class Solution {
    public int solution(int[] topping) {
        int answer = 0;
        
        Map<Integer, Integer> map_1 = new HashMap<>(); // 형
        Map<Integer, Integer> map_2 = new HashMap<>(); // 동생
        
        for (int e : topping) {
            map_2.put(e, map_2.getOrDefault(e, 0) + 1);
        }
        
        for (int e : topping) {
            map_1.put(e, map_1.getOrDefault(e, 0) + 1);
            
            if (map_2.get(e) - 1 == 0)
                map_2.remove(e);
            else
                map_2.put(e, map_2.get(e) - 1);  
            
            if (map_1.size() == map_2.size())
                answer++;
        }

        return answer;
    }
}

File: woojae/2024_5/0504/PGS_롤케이크 자르기.py
def solution(topping):
    chulsu, brother = Counter(topping), set()
    answer = 0

    for i in topping:  # 케이크 자르기는 순차적으로 진행됨
        chulsu[i] -= 1  # 순서대로 토핑 제거
        brother.add(i)  # 제거한 토핑 추가

        if not chulsu[i]:  # 더 이상 제거할 토핑이 없다면
            chulsu.pop(i)  # 해당 토핑 완전 제거
        if len(chulsu) == len(brother):  # 형과 동생이 가진 토핑의 종류가 같다면
            answer += 1  # 케이크를 자르는 방법의 수 추가

    return answer
